structure change for Chinese version

lisa0314 opened the pull request:
1. The configuration about the choices of the layout templates is in the _data.json in each directory by adding some
   language property
2. the _layout templates for Chinese and English Version are separated
3. there exist different files about navigation column for different languages
4. all files for Chinese Version are suffixed with "_zh"

lisa0314 commented:
@jondong

jondong commented:
LGTM, thanks for the PR!
@bspencer would you please help to review? Thanks

bspencer commented:
OK, I think the strategy will work, although it does duplicate every file and probably doesn't scale well to lots of languages.   A few notes before we merge:
 1) the selection dropdown is ugly.  See https://software.intel.com/en-us/xdk/docs/why-use-crosswalk-for-android-builds for a cleaner example (bottom-right).   We need to replace the "get started" button in the header.  The drop-down must fit on one row.  The home page is esp bad because it adds a whole new row. The drop-down should be subtle, in the same font and color as the other menu items.  Here is how the pull looks now:  http://i.imgur.com/tdd3sHE.png   Here is how I'd like it to look: http://imgur.com/TIcykic
2) When I click on the documentation menu items with sub-topics (e.g. iOS), the tree expands partially but the menu items don't show up.
3) When I click "chinese" on the home page, I get "404 not found".

lisa0314 commented:
@bspencer , Thanks for your review. 
1)The first note you listed may is the older version of the branch I have submitted few days ago.  The newest commit has removed "get started" button and showed a new selection dropdown for language change. The newest commit was unique as 9727f2e.
2) For the second note, I just committed the main structure change at the moment. The reason for why the menu items don't show up is that I have not committed the _data.json file for the second-level directory, such as document/ios, document/android.
3) The reason for the third problem is that I have just committed the configuration files not including the index file. This commit was aimed for stating the basic principles of how to add the Chinese version. I will commit the relevant files for the content of the website.
Thank you!

lisa0314 commented:
@jondong

bspencer commented:
1)The first note you listed may is the older version of the branch I have submitted few days ago.

True!  Very sorry.  I pulled again and refreshed the browser which had the stale content.  It looks much better.  I’d still like the look I put in the comment with the same font, not bolded, not standing out.  Also, the drop-down menu hides behind the text on the home page for some reason, but you probably already knew that.  Also, the names should always be capital first letter (English, not english).
Thanks

From: lisa0314 [mailto:notifications@github.com]
Sent: Monday, April 18, 2016 9:15 PM
To: crosswalk-project/crosswalk-website
Cc: Spencer, Bob
Subject: Re: [crosswalk-project/crosswalk-website] structure change for Chinese version (#293)


@bspencer<https://github.com/bspencer> , Thanks for your review.
1)The first note you listed may is the older version of the branch I have submitted few days ago. The newest commit has removed "get started" button and showed a new selection dropdown for language change. The newest commit was unique as 9727f2e<https://github.com/crosswalk-project/crosswalk-website/commit/9727f2e210851b9feb89c6f6be582c1b395ec6a3>.
2) For the second note, I just committed the main structure change at the moment. The reason for why the menu items don't show up is that I have not committed the _data.json file for the second-level directory, such as document/ios, document/android.
3) The reason for the third problem is that I have just committed the configuration files not including the index file. This commit was aimed for stating the basic principles of how to add the Chinese version. I will commit the relevant files for the content of the website.
Thank you!

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/crosswalk-project/crosswalk-website/pull/293#issuecomment-211708035>

lisa0314 commented:
@bspencer , thanks for your review. I have committed a new patch.
1)Firstly, according to your review comments, I have modified the interface of the button changing the language version. 
2)Besides, I also added the index file for Chinese version of each directory in the newest commit which is benefit for demonstration at the moment. For instance, when you click "chinese" on the home page now, you could see the home page of Chinese version.
3)For the demonstrate of document section and the navigation column, you can try the section of contribution at present. This commit is just aimed to demonstrate the structure change for Chinese version. With the section of document containing so many files, I have not add these translation files in the newest commit. However, I will commit the translation soon after this commit is merged.

lisa0314 commented:
@jondong

bspencer commented on public/_includes/documentation-nav.ejs:
```diff
@@ -15,7 +16,7 @@ function endsWith(str, suffix) {
        for(var item1 in public.documentation._data) { 
          var article1 = public.documentation._data[item1];
-
+         if ( article1.language === "en" ) {
```
For navigation, is it possible to use one set of files instead of 2 duplicate files?  The only difference between this and the _zh one is that everywhere it says "if (article1.language === "zh")...."   For the content, it makes sense to have duplicate files because they are translations, but for the navigation and rules such as this, can we design it to have one set of files that is language aware?

bspencer commented on public/_includes/documentation-next.ejs:
```diff
@@ -5,7 +5,7 @@
     <% var item = subpages._data[slug] %>
     <% if(slug === current.source) { %>
       <% var next = true %>
-    <% } else if(next === true) { %>
```
Same comment here -- let's merge the files with language logic inside.

bspencer commented on public/documentation/_data.json:
```diff
@@ -1,50 +1,108 @@
 {
   "about": {
-    "title": "About"
```
Now I'm going to jump to the other argument for _data.json...  Is there a reason to have just a single _data.json instead of _data_en.json, _data_zh.json, ...  :)

bspencer commented:
If we can get the logic working without any UI, then I can merge those changes.  Otherwise, if we have a UI then there are more things we have to do before I can merge.  The main question I have is what is the plan to gracefully fallback if there is no "zh" translation for a given page?  Does it tell the user "There is no Chinese translation for this page." or what do you want to do?

Thanks for the changes to the menu.  There are still the following issues with the UI:
  - It needs to behave good when the window is resized (smaller)
  - The drop-down should not be covered by other words (you can see when the screen is smaller)
But... you don't need to worry about these UI issues for now.  We can work on the backend logic first  (nav.ejs, documentation-nav_zh.ejs, and _data.json)

Once more Q:  do we have the rights to display the Weibo graphic?  For twitter they have a public graphic that can be legally used.  Not sure what the rules are for Weibo.

jondong commented:
@bspencer Thanks for the review and comments, we are working on a new solution to address your comments, the basic concepts are:
1. we'll keep the template file as one (e.g. nav.ejs) instead of duplicate them by languages. And as template file also contain contents which needs to be translated, we are going to extract them into the language data files (e.g. _nav_en.json, _nav_zh.json), and load them in the template file and generate contents according to the language selection.
2. We are going to split the data files by language, (e.g. _data.json and _data_zh.json), for the page which does not have the current language translation, we will fallback to show the english contents, with a notice toaster to let user aware the translation is not available.

The proposed patch is going to be submitted soon by Lisha, any thoughts from u please feel free to let us know.

For the Weibo VI, the picture is downloaded from Weibo's branding logo publish page: http://weibo.com/static/logo?bottomnav=1&wvr=6 and it is intended to be legally used by the third party like us.

lisa0314 commented:
@bspencer  Thanks for your review. 
1. According to your comments, we have merge the separated template files for English and Chinese version together, for instance, there exists only one _nav.ejs file.The contents which need to be translation, we have extracted them into a data file in public/content/, and load them into the template file and generate contents  according to the language selection.
2. As for the argument for _data.json, I try to separate it into two files according to language property, just have two files _data_en.json, _data_zh.json instead of one single file. However, I found out that the harp only support one _data.json file in each directory and the rename of the file was not identified such as _data_zh.json. Maybe I have missed some features of harp, so please share with me if you know the solution for this.
3. We haven't come up with a nice solution for the fallback problem if there is no "zh" translation for a given page. Could you pleased give some suggestions?

lisa0314 commented:
@bspencer  Thanks for your review.
1. According to your comments, we have merge the separated template files for English and Chinese version together, for instance, there exists only one _nav.ejs file.The contents which need to be translation, we have extracted them into a data file in public/content/, and load them into the template file and generate contents according to the language selection.
2. As for the argument for _data.json, I try to separate it into two files according to language property, just have two files _data_en.json, _data_zh.json instead of one single file. However, I found out that the harp only support one _data.json file in each directory and the rename of the file was not identified such as _data_zh.json. Maybe I have missed some features of harp, so please share with me if you know the solution for this.
3. For the fallback problem if there is no "zh" translation for a given page, we regard it as ‘Page Not Found'.  We will show the user a new page with Chinese navigation column and content when faced with fallback problem. Then the user can browse the site of Chinese version by clicking the navigation bar or go back to the English version.  To implement this solution, we have removed the file named '404.md' , added '404.ejs' and '404_zh.ejs' instead. In the file of 404.ejs, we use the value of cookie to decide if there needs to a redirection.

lisa0314 commented:
@jondong

jondong commented on public/_includes/documentation-nav.ejs:
```diff
@@ -12,62 +13,72 @@ function endsWith(str, suffix) {
          var pageRef = "/documentation/" + item1.toLowerCase() + (endsWith (item1, ".php") ? "" : ".html"); %>
        <li class="article-item <%= isFile ? ' nav-item--active' : '' %>">
-         <a class="article-link" href="<%= pageRef %>"><%= article1.title %></a>
+         <a class="article-link" href="<%= pageRef %>"><%= article1.title  %></a>
```
extra whitespace which is not needed.

jondong commented on public/_includes/documentation-next.ejs:
```diff
@@ -1,14 +1,16 @@
 <% var subpages = public.documentation[current.path[1]] %>
+<% var object = public.content._data; %>
```
this could be a more meaningful name than "object"

jondong commented on public/_includes/foot.ejs:
```diff
@@ -0,0 +1,51 @@
+<%
+   var object = public.content._data;
```
this could be a more meaningful name than "object"

jondong commented on public/_includes/foot.ejs:
```diff
@@ -0,0 +1,51 @@
+<%
+   var object = public.content._data;
+   var lang = ( current.source.lastIndexOf("_zh") > 0 ? "zh": "en");
```
extra whitespace

jondong commented on public/_includes/nav.ejs:
```diff
@@ -1,60 +1,88 @@
-<nav class="nav">
-  <a href="/" class="nav-logo logo">
+<%		
+   var object = public.content._data;
```
ditto

jondong commented on public/_includes/nav.ejs:
```diff
@@ -1,60 +1,88 @@
-  <a href="/" class="nav-logo logo">
+<%		
+   var object = public.content._data;
+   var lang = ( current.source.lastIndexOf("_zh") > 0 ? "zh": "en");
```
extra whitespace

jondong commented on public/contribute/_nest.ejs:
```diff
@@ -1,12 +1,18 @@
+<%
+   var object = public.content._data;
+   var lang = ( current.source.lastIndexOf("_zh") > 0 ? "zh": "en");
```
ditto

jondong commented on public/contribute/_nest.ejs:
```diff
@@ -1,12 +1,18 @@
-      <li class="article-item <%= current.source === slug.toLowerCase() ? ' nav-item--active' : '' %>">
-        <a href="/contribute/<%= slug ? slug.toLowerCase() + '.html' : 'index.html' %>"><%= article.title %></a>
-      </li>
+      <% if( article.language == lang) { %>
```
if (article...

jondong commented on public/contribute/index.ejs:
```diff
@@ -11,9 +11,11 @@
-          <li class="article-item <%= current.source === slug.toLowerCase() ? ' nav-item--active' : '' %>">
-            <a href="/contribute/<%= slug ? slug.toLowerCase() + '.html' : '' %>"><%= article.title %></a>
-          </li>
+          <% if( article.language === "en") { %>
```
ditto

jondong commented on public/documentation/getting_started_zh.md:
```diff
@@ -0,0 +1,108 @@
+# 开始使用Crosswalk项目
+
+欢迎使用Crosswalk项目，对于你的移动和桌面web应用最好的运行时。选择下方你感兴趣的部分然后点击进入！
```
用于构建移动和桌面应用最好的web运行时环境。

jondong commented on public/documentation/getting_started_zh.md:
```diff
@@ -0,0 +1,108 @@
+
+欢迎使用Crosswalk项目，对于你的移动和桌面web应用最好的运行时。选择下方你感兴趣的部分然后点击进入！
+
+<h2>Crosswalk相关开发</h2>
```
Crosswalk开发

jondong commented on public/documentation/getting_started_zh.md:
```diff
@@ -0,0 +1,108 @@
+欢迎使用Crosswalk项目，对于你的移动和桌面web应用最好的运行时。选择下方你感兴趣的部分然后点击进入！
+
+<h2>Crosswalk相关开发</h2>
+<p>搭建你的开发系统，然后开始使用Crosswalk项目的运行时和Web视图来创建应用。</p>
```
搭建你的开发环境，然后开始使用Crosswalk运行时环境和Webview来创建应用

jondong commented on public/documentation/getting_started_zh.md:
```diff
@@ -0,0 +1,108 @@
+
+<div class="featureBlock"> 
+<!--
+  <div class="featureDiv" style="background: rgba(224,224,224, 0.4);">
```
let's delete those codes in comments.

jondong commented on public/documentation/getting_started_zh.md:
```diff
@@ -0,0 +1,108 @@
+      <div class="featureHeader"><img class="featureImg" src="/assets/icons/logo-android-white.png"/><div class="featureTitle">Android target</div></div>
+      <hr class="featureSplitter" />
+      <div class="featureBody">
+        为Android平台创建应用
```
我们统一把Android翻译成安卓吧。

jondong commented on public/documentation/_data.json:
```diff
@@ -1,50 +1,146 @@
+    "origin": "samples"
+  },
+  "community_zh": {
+    "title": "社团",
```
社区

jondong commented on public/documentation/_data.json:
```diff
@@ -1,50 +1,146 @@
+    "origin": "samples"
+  },
+  "qa_zh": {
+    "title": "质量保证",
```
这里面直接用QA比较合适

jondong commented on public/documentation/android/_data.json:
```diff
@@ -1,26 +1,76 @@
+    "language": "en"
+  },
+  "system_setup_zh": {
+    "title": "系统创建",
```
系统设置

jondong commented on public/documentation/getting_started_zh.md:
```diff
@@ -0,0 +1,108 @@
+  <br clear="all" />
+</div>
+<h2>项目功能</h2>
+<p>了解使得Crosswalk项目成为世界水平的其他部分</p>
```
其它让Crosswalk成为世界级项目的内容，来了解一下：

jondong commented on public/documentation/windows/_data.json:
```diff
@@ -1,26 +1,74 @@
+    "language": "en"
+  },
+  "windows_host_setup_zh": {
+    "title": "Windows主机搭建",
```
Windows开发环境设置

jondong commented on public/documentation/windows/_data.json:
```diff
@@ -1,26 +1,74 @@
+    "layout": "../../_layout_zh"
+  },
+  "deploy_windows_store_zh": {
+    "title": "部署到Windows store",
```
Windows Store

jondong commented on public/documentation/windows/_data.json:
```diff
@@ -1,26 +1,74 @@
+    "layout": "../../_layout_zh"
+  },
+  "build_crosswalk_for_windows_zh": {
+    "title": "为windows平台创建Crosswalk",
```
在Windows上编译Crosswalk
